Conjur/conventions/ruby-style-guide.md

```diff
@@ -24,10 +23,194 @@ This guide is intended to be a "living document" in the sense that it's expected
+
+- Avoid "name stutter". That is, don't repeat a name from the module path in a class, and don't repeat a name from the class in a method. For example:
+
+   ```ruby
+   # bad
+   class MyCar
+     def start_car_engine
+       # ...
+     end
+   end
+   
+   # good
+   class MyCar
+     def start_engine
+       # ...
+     end
+   end
+   
+   # bad
+   module Authenticators
+     class OktaAuthenticator
+       #...
+     end
+   end
+   
+   # good
+   module Authenticators
+     class Okta
+       #...
+     end
+   end
+   ```
+
+- In command class we should split methods called in `call` function into 2 parts - one with an explicit name so that the command class 'call' function tells the story, and one with the returned object name for memoization.
+
+  - Bad:
+
+    ```ruby
+    def get_jwt_identity_from_request
+        @logger.debug(LogMessages::Authentication::AuthnJwt::CallingGetJwtIdentity.new)
+        @jwt_identity ||= @jwt_configuration.jwt_identity
+        @logger.info(LogMessages::Authentication::AuthnJwt::FoundJwtIdentity.new(jwt_identity))
+    end
+    ```
+
+  - Good:
+
+    ```ruby
+    def get_jwt_identity_from_request
+        @logger.debug(LogMessages::Authentication::AuthnJwt::CallingGetJwtIdentity.new)
+        jwt_identity
+        @logger.info(LogMessages::Authentication::AuthnJwt::FoundJwtIdentity.new(jwt_identity))
+    end
+    
+    def jwt_identity
+        @jwt_identity ||= @jwt_configuration.jwt_identity
+    end
+    ```
+
+- In constuctor of classes separate input handling and dependency handling with empty line
+
+- In command class implement methods in the order they are called
+
+  ```ruby
+  def call
+    func_a
+    func_b
+    func_c
+  end
+  
+  private
+  
+  def func_a
+    ...
+  end
+  
+  def func_b
+    ...
+  end
+  
+  def sub_func_b
+    ...
+  end
+  
+  def func_c
+    ...
+    end
+  ```
+
+- In functions with many parameters put each parameter in seperate line in the declaration. In this case end each param with ":" so the caller will need to specify name of variable for each param.
+
+  ```
+  def method(
+    param1:
+    param2:
+    param3:
+  )
+  ```
+
+  
+
+- When overriding a code climate violation with an inline exception (this isn't possible for all code climate  plugins but is for many), add the reason for the exception on the line above the special "ignore" directive
+
+   https://github.com/cyberark/conjur/blob/5fe6b22f225a1f2cf0be90d90f184c74c150be5a/cucumber/policy/features/support/rest_client_wrapper.rb#L12
+
+## Design Baseline
+
+- On any new class ask yourself if it should be a class or command class. Command Classes coordinate the execution of a multi-step process.
+
+  https://github.com/cyberark/conjur/blob/master/app/domain/README.md#the-functional-object-oriented-style
+
+  
+
+- No need for interface classes
+
+- *Never* use inheritance or mixins.
+
+- Dependencies for classes should be passed as default values in constructor
```

Comment:
```suggestion
- Dependencies for classes should be passed in the constructor
```
I'm seeing a misunderstanding of dependency injection which is easy to miss with the Command class.  All the command class does is provide a nice DSL over vanilla Ruby:

```ruby
module Authenticators
  class Okta
    def initialize(logger:, configuration:)
      @logger = logger
      @configuration = configuration
    end

    def call(request:)
      ....
    end
  end
end
```

vs

```ruby
module Authenticators
  class Okta
    extend CommandClass::Include

    command_class(
      dependencies: {logger:, configuration:},
      inputs: %i[request]
    ) do

      def call(request:)
        ....
      end
    end
  end
end  
```

In the top example, the initialization arguments depend on duck-typing (the interface of the dependency matters).  @jonahx's Command class reduces the amount of boilerplate code, but (IMO), reduces the transparency of exactly how an injected dependency is injected (and whether a dependency should be injected into the `initialize` or `call` method).

I'd love to see us moving injection up the stack, so:

```ruby 
Authentication::AuthnOidc::UpdateInputWithUsernameFromIdToken.new.(
  authenticator_input: authenticator_input
)
```

becomes:

```ruby
Authentication::AuthnOidc::UpdateInputWithUsernameFromIdToken.new(
  fetch_authenticator_secrets: Authentication::Util::FetchAuthenticatorSecrets.new(
    fetch_secrets: ::Conjur::FetchRequiredSecrets.new(
      resource_class: ::Resource
    )
  ),
  validate_account_exists: ::Authentication::Security::ValidateAccountExists.new(
    role_class: ::Role
  ),
  verify_and_decode_token: ::Authentication::OAuth::VerifyAndDecodeToken.new(
    fetch_provider_keys: ::Util::ConcurrencyLimitedCache.new(
      ::Util::RateLimitedCache.new(
        FetchProviderKeys.new(
          logger: logger,
          discover_identity_provider: DiscoverIdentityProvider.new(
            logger: logger,
            open_id_discovery_service: OpenIDConnect::Discovery::Provider::Config
          )
        ),
        refreshes_per_interval: 10,
        rate_limit_interval: 300, # 300 seconds (every 5 mins)
        logger: logger
      ),
      max_concurrent_requests: 3, # TODO: Should be dynamic calculation
      logger: logger
    ),
    verify_and_decode_token: ::Authentication::Jwt::VerifyAndDecodeToken.new(
      jwt_decoder: JWT,
      logger: logger
    ),
    logger: logger
  ),
  logger: logger
).call(
  authenticator_input: authenticator_input
)
```

Although the second option is substantially more verbose, it's very clear what the dependencies are.  We can clearly see the OIDC authenticator uses the Conjur `Role` and `Resource`, something that was not obvious when default values were used int the initialization block.

Reply:
Im not sure if its really readable . 
logger: logger repeates it self so many times in a single place its little bit strange to me.

Reply:
@tzheleznyak, exactly the point of being explicit. In the current implementation, the method call looks very simple and straightforward.  In truth, it's anything but. In reality, the `UpdateInputWithUsernameFromIdToken` initializer has 23 dependencies (in the entire tree). 

If this complexity is unsatisfactory, we should deal with the root cause of the complexity. Maybe we move Policy loading outside the authenticator, maybe we refactor the two implementations of `VerifyAndDecodeToken` to use a single class rather than two seperate (`Jwt` and  `OAuth`). The test is the time it takes a new engineer to come up to speed, the seemingly "simple" top level call that hide the complexity. 

IMO, the number of security issues stemming from authenticators can be traced to their complexity.  Simple to understand code is far less likely to have bugs.